### PR TITLE
ACCUMULO-4669 Use windowed statistics in RFile

### PR DESCRIPTION
**Pull request by keith-turner**

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java [hunk 1]
```diff
@@ -416,8 +416,9 @@ public void flushIfNeeded() throws IOException {
     }
 
     private boolean isGiantKey(Key k) {
-      // consider a key thats more than 3 standard deviations from previously seen key sizes as giant
-      return k.getSize() > keyLenStats.getMean() + keyLenStats.getStandardDeviation() * 3;
+      double mean = keyLenStats.getMean();
+      double stddev = keyLenStats.getStandardDeviation();
+      return k.getSize() > mean + 3 * Math.max(mean, stddev);
```
Comment (mikewalch) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java]: Could add some comments here about why this logic is used to determine if key is giant...

File: core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java [hunk 2]
```diff
@@ -403,7 +402,8 @@ public void flushIfNeeded() throws IOException {
 
     private SampleLocalityGroupWriter sample;
 
-    private SummaryStatistics keyLenStats = new SummaryStatistics();
+    // Use windowed stats to fix ACCUMULO-4669
+    private RollingStats keyLenStats = new RollingStats(2017);
```
Comment (mikewalch) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java]: Why 2017 for window size?

File: core/src/main/java/org/apache/accumulo/core/file/rfile/RollingStats.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  RollingStats(int windowSize) {
+    this.windowFull = false;
+    this.position = 0;
+    this.window = new double[windowSize];
+  }
+
+  /**
+   * @see <a href= "http://jonisalonen.com/2014/efficient-and-accurate-rolling-standard-deviation/">Efficient and accurate rolling standard deviation</a>
+   */
+  private void update(double n, double o, int w) {
```
Comment (mikewalch) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/RollingStats.java]: I guess `n` & `o` is for new & old.  Could instead use `newValue` & `oldValue` to make things clear.  What is `w`?

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

**Comment by keith-turner** [comment]
merged in 26e83f05d1448631f0b8a0da1b8671abe9beb922

**Pull request closed without merging**